### PR TITLE
environment variables support for timeout

### PR DESCRIPTION
**Pull request by alonisser**
Signed-off-by: alonisser <alonisser@gmail.com>

## Related Issues/PRs

#6970 

## What changes are proposed in this pull request?

Add environment variable under mlflow.environment_variables and reference it in the scoring server

## How is this patch tested?

Not sure on how to test this, I've looked at the current tests and I didn't see a relevant example. 

## Does this PR change the documentation?

- [ X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.
Allow user to configure via environment variable in the deployed mlflow model the scoring server timeout

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [X] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/environment_variables.py
Comment (dbczumar) [inline mlflow/environment_variables.py:174]:
```suggestion
#: Specifies the MLflow Model Scoring server request timeout in seconds
```

**Review by dbczumar — COMMENTED** [review]
@alonisser This is fantastic! A few small questions:

1. Can we replace https://github.com/mlflow/mlflow/blob/96f861b9f32a8e6ad2473db3b3272a50f5f8e71a/mlflow/models/cli.py#L31 with 

```
click.option(
    "--timeout", "-t", help="Timeout in seconds to serve a model scoring request (default: 60).",
    envvar="MLFLOW_SCORING_SERVER_REQUEST_TIMEOUT",
)
```
? This way, the MLflow CLI will clearly document that it respects the timeout environment variable.

2. Do we need to set this environment variable from within the Pandas UDF created by `mlflow.pyfunc.spark_udf`? (https://github.com/mlflow/mlflow/blob/96f861b9f32a8e6ad2473db3b3272a50f5f8e71a/mlflow/pyfunc/__init__.py#L1030-L1032)

As far as I'm aware, these environment variables from the driver aren't automatically propagated to the UDF.

3. Should we define a similar environment variable for the scoring server readiness check timeout? https://github.com/mlflow/mlflow/blob/b08aa03ed35f76ee143022ac2d535576fecfaeb2/mlflow/pyfunc/__init__.py#L584

File: mlflow/environment_variables.py
Comment (alonisser) [reply mlflow/environment_variables.py:174]: Thanks . would do

**Comment by alonisser** [comment]
1. I'll check if that propagates to the same var, not sure
2. I'm not aware of that, although I thought anyway we'd need to setup environment variables in all nodes and not just in the driver, is there a place where the udf environment is built? can you please point to where?
3. I'm not aware of this need, is there a need? Should I add it for completeness' sake?

**Comment by alonisser** [comment]
About 1, I see it uses 
```python
        server_implementation = mlserver if enable_mlserver else scoring_server

```
What is mlserver and should I enable this change there also? or is it a stand alone package?

**Comment by alonisser** [comment]
Pushing a fix for 1 , I understand now as I fixed that I need to change the place the timeout is checked vs envvars to support mlserver also, so changed it . still investigating 2, I'd be happy for pointers if possible

File: mlflow/environment_variables.py
Comment (alonisser) [reply mlflow/environment_variables.py]: fixed

**Comment by alonisser** [comment]
Following 2, As far as a I see the code, it's either handled magically by spark (probably not) or we need to setup environment variables in the executors too for that to work (What Is probably what we'll need to do anyway), it can't be handled from inside mlflow creating the udf afaik

Also following 2. I see my changes about scoring server didn't influence the actual scoring server "inside" the udf, so I've implemented the change there also 

Appreciate the help!

**Comment by alonisser** [comment]
About 2, I see the code takes care of copying environment variables inside the subprocess process

https://github.com/mlflow/mlflow/blob/96f861b9f32a8e6ad2473db3b3272a50f5f8e71a/mlflow/pyfunc/scoring_server/__init__.py#L398

**Comment by dbczumar** [comment]
> About 2, I see the code takes care of copying environment variables inside the subprocess process
> 
> https://github.com/mlflow/mlflow/blob/96f861b9f32a8e6ad2473db3b3272a50f5f8e71a/mlflow/pyfunc/scoring_server/__init__.py#L398

Awesome! Great find!

File: mlflow/utils/cli_args.py
Comment (dbczumar) [inline mlflow/utils/cli_args.py:161]: This timeout is used in other CLIs such as `mlflow server`, which are not related to Model serving. Can we instead make the change in https://github.com/mlflow/mlflow/blob/96f861b9f32a8e6ad2473db3b3272a50f5f8e71a/mlflow/models/cli.py#L31?

**Review by dbczumar — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/6988/files#r991708001 is addressed. I think we can hold off on the scoring server setup timeout for now. Thanks @alonisser!

File: mlflow/utils/cli_args.py
Comment (alonisser) [reply mlflow/utils/cli_args.py:161]: It's the same one I believe :) . `cli_args.TIMEOUT` references the other one. and semantically it's all the same "timeout for the scoring server"

But note, since this isn't passed into the spark_udf implementation, I've added a direct reference to the env var there also

File: mlflow/pyfunc/scoring_server/__init__.py
Comment (dbczumar) [inline mlflow/pyfunc/scoring_server/__init__.py:368]: I think the following edit is necessary since `timeout` is `None` by default, which would break line 370:

```
timeout = timeout or MLFLOW_SCORING_SERVER_REQUEST_TIMEOUT.get()
```

File: mlflow/utils/cli_args.py
Comment (dbczumar) [reply mlflow/utils/cli_args.py:161]: My mistake - this is, indeed the same arg. Thanks @alonisser !

**Review by dbczumar — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/6988/files#r993997278 is addressed. Thanks so much, @alonisser !

File: mlflow/pyfunc/scoring_server/__init__.py
Comment (alonisser) [reply mlflow/pyfunc/scoring_server/__init__.py:368]: I've removed that because the code calling it is either from 
evaluate for model
https://github.com/zencity/mlflow/blob/6f2231b3cd4be7ad8558d4a61e4220668e2529c6/mlflow/pyfunc/__init__.py#L576 (yes I've found another entry point to the scoring server)

Or the inner pyspark udf 

https://github.com/zencity/mlflow/blob/6f2231b3cd4be7ad8558d4a61e4220668e2529c6/mlflow/pyfunc/__init__.py#L1080

Or from

https://github.com/zencity/mlflow/blob/6f2231b3cd4be7ad8558d4a61e4220668e2529c6/mlflow/pyfunc/backend.py#L162

I'd rather keep it outside, so we'll support future additions to other scoring backends, mlserver supporting timeout (currently doesn't) etc. 

if not, let's put None, in all of those places, and add the line you suggested, but I don't think it's future proof

Comment (alonisser) [reply mlflow/pyfunc/scoring_server/__init__.py:368]: @dbczumar

Comment (dbczumar) [reply mlflow/pyfunc/scoring_server/__init__.py:368]: Hi @alonisser , I'd rather put None in other places and put the line I suggested here. The reason is that, otherwise, if this method is called with the default `timeout=None`, it will break. It should not be the responsibility of each call site to supply a valid value of this argument else face an error, since this argument is optional.

**Comment by alonisser** [comment]
Got it. Will do

בתאריך יום ו׳, 14 באוק׳ 2022, 7:39, מאת Corey Zumar ‏<
***@***.***>:

File: mlflow/pyfunc/scoring_server/__init__.py
Comment (alonisser) [reply mlflow/pyfunc/scoring_server/__init__.py:368]: Done, I've kept two places also in the calling sites (removed one) as the original was hard coded to 60 and reverting to that is problematic. So we're both explicit in the calling site and ensured inside the scoring server logic @dbczumar

Comment (alonisser) [reply mlflow/pyfunc/scoring_server/__init__.py:368]: @dbczumar

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks so much, @alonisser !

**Comment by alonisser** [comment]
So @dbczumar I think the added tests exposed an issue with mlserver with my changes, I'll explore it
But some tests failing seems to be unrelated to my changes, can you please verify if this is a recurring issue and a test problem or related to this PR?

https://github.com/mlflow/mlflow/actions/runs/3242324978/jobs/5374541707#step:7:916

**Comment by alonisser** [comment]
Well I've looked into it

While the second ("models") failure is harder to understand (with gunicorn wrapped in docker) and still trying to recreate. The first one is because of the mlserver "scoring" backend not supporting timeout. and throwing an exception (in mlflow code, not mlserver) about not supporting it. Now that I've added the default value it gets the timeout and throws

There are two options
1. Remove the default value (while keeping it "documented" as it was before) and add it in another place inside the scoring server as a last line of defence default (as it can't get None) 

https://github.com/zencity/mlflow/blob/6f2231b3cd4be7ad8558d4a61e4220668e2529c6/mlflow/utils/cli_args.py#L162

Which in my view isn't great, as documentation becomes misleading, and add 60 in multiple places hard coded

2. remove the exception and just log and error and delete the "timeout" argument if passed

WDYT @dbczumar

**Comment by dbczumar** [comment]
> Well I've looked into it
> 
> While the second ("models") failure is harder to understand (with gunicorn wrapped in docker) and still trying to recreate. The first one is because of the mlserver "scoring" backend not supporting timeout. and throwing an exception (in mlflow code, not mlserver) about not supporting it. Now that I've added the default value it gets the timeout and throws
> 
> There are two options
> 
> 1. Remove the default value (while keeping it "documented" as it was before) and add it in another place inside the scoring server as a last line of defence default (as it can't get None)
> 
> https://github.com/zencity/mlflow/blob/6f2231b3cd4be7ad8558d4a61e4220668e2529c6/mlflow/utils/cli_args.py#L162
> 
> Which in my view isn't great, as documentation becomes misleading, and add 60 in multiple places hard coded
> 
> 2. remove the exception and just log and error and delete the "timeout" argument if passed
> 
> WDYT @dbczumar

Hi @alonisser, thank you for investigating! I agree that (2) is the best option :)

**Comment by alonisser** [comment]
@dbczumar I've pushed the discussed fix, I believe It would fix "MLflow tests / python (pull_request)" so I'd appreciate if you approve a rerun to validate that , what I'm wondering is about "MLflow tests / models (pull_request) " Which I couldn't recreate locally, if they would still fail, I'd need help to setup this locally, as those tests currently fail locally on completely different errors then in the CI

**Comment by alonisser** [comment]
Also I've intergrated the changes merged into master since I've forked

**Comment by alonisser** [comment]
@dbczumar I've managed to run the previously failed tests locally and they pass now. The problem is I don't think they would work in the CI, due to this commit being merged yesterday

https://github.com/mlflow/mlflow/commit/f76de49b49142a2f41eb478b3f30807fc1cbbff7#diff-535d7991b2be4fed9e50b08954351ac640d5be083e8ca0bae614a78f44e66e9bR5

the docker build image command (being tested there) tries to do a pip install of mlflow with 2.0.0dev version. but this version doesn't exist in PYPI yet. Not something I can solve from "my side". Can't figure out if any new PR would actually pass this until it's fixed

**Comment by dbczumar** [comment]
> @dbczumar I've managed to run the previously failed tests locally and they pass now. The problem is I don't think they would work in the CI, due to this commit being merged yesterday
> 
> [f76de49#diff-535d7991b2be4fed9e50b08954351ac640d5be083e8ca0bae614a78f44e66e9bR5](https://github.com/mlflow/mlflow/commit/f76de49b49142a2f41eb478b3f30807fc1cbbff7#diff-535d7991b2be4fed9e50b08954351ac640d5be083e8ca0bae614a78f44e66e9bR5)
> 
> the docker build image command (being tested there) tries to do a pip install of mlflow with 2.0.0dev version. but this version doesn't exist in PYPI yet. Not something I can solve from "my side". Can't figure out if any new PR would actually pass this until it's fixed

Hi @alonisser, I've re-triggered tests against this PR and plan to merge unless there's a clear connection between any failures and the PR (I'd imagine any failures will be completely unrelated, as you've pointed out). Thanks for bearing with us and following through with this excellent contribution!

**Comment by dbczumar** [comment]
> Thanks @dbczumar ! but I see there is a test checking the exception I've removed
> 
> FAILED tests/pyfunc/test_mlserver.py::test_raise_error_mlserver_timeout[params0-Timeout is not yet supported in MLServer.]
> 
> So please let me first remove this test and then can be merged - would it be preferred to keep the test and skip it or just remove it?

I think removing the test is perfect! :)

**Comment by alonisser** [comment]
> > Thanks @dbczumar ! but I see there is a test checking the exception I've removed
> > FAILED tests/pyfunc/test_mlserver.py::test_raise_error_mlserver_timeout[params0-Timeout is not yet supported in MLServer.]
> > So please let me first remove this test and then can be merged - would it be preferred to keep the test and skip it or just remove it?
> 
> I think removing the test is perfect! :)

Done. thanks

**Comment by dbczumar** [comment]
> > > Thanks @dbczumar ! but I see there is a test checking the exception I've removed
> > > FAILED tests/pyfunc/test_mlserver.py::test_raise_error_mlserver_timeout[params0-Timeout is not yet supported in MLServer.]
> > > So please let me first remove this test and then can be merged - would it be preferred to keep the test and skip it or just remove it?
> > 
> > 
> > I think removing the test is perfect! :)
> 
> Done. thanks

Awesome! Let me know once you've pushed the changes and I can re-trigger tests :)

**Comment by alonisser** [comment]
Sorry @dbczumar forgot to push before. pushed now.

**Comment by alonisser** [comment]
Rebased with latest changes from repo, frankly I don't know thy the R test is failing, and don't know how to check that and how is the R server actually related to my changes

**Comment by dbczumar** [comment]
> Rebased with latest changes from repo, frankly I don't know thy the R test is failing, and don't know how to check that and how is the R server actually related to my changes

At a glance, I'm not sure either. I'll dig in and get this fixed very shortly :)

**Comment by dbczumar** [comment]
Hi @alonisser , I've filed https://github.com/zencity/mlflow/pull/1 against your PR branch, which should resolve the issue. If it looks okay to you, can you merge it? Thanks so much!

**Comment by dbczumar** [comment]
> > Hi @alonisser , I've filed [zencity#1](https://github.com/zencity/mlflow/pull/1) against your PR branch, which should resolve the issue. If it looks okay to you, can you merge it? Thanks so much!
> 
> Wow thanks! , I've merged

Thanks so much, @alonisser ! Awesome PR! Merging! :)

**Comment by alonisser** [comment]
Hey, How do I know when this would be released? What's the lifecycle of releases on the mlflow project?

**Pull request merged**
